### PR TITLE
Use `DeviceBuffer` in `device_array`

### PR DESCRIPTION
**Pull request by jakirkham**
Instead of relying on using `rmm_alloc`, this uses `DeviceBuffer` to allocate memory to back Numba `DeviceNDArray` objects. This has the benefit that the memory will be freed (by C++) once the `DeviceBuffer` object scopes out. So the finalizer is no longer needed/used here. Have dropped the `_array_helper` function. `_make_finalizer` is kept as downstream projects (cuDF and cuGRAPH) still use it, but is no longer used by RMM.

Note: This is targeting RAPIDS 0.13.0. Making it a draft now for discussion.

### CODE REVIEW

**Comment by jrhemstad** [comment]
Alternatively, consider that C++ has a `rmm::device_vector` class that I think would be a more natural backing for `device_array`.

**Comment by jakirkham** [comment]
Interesting. Thanks for mentioning that. What are the main differences between the two?

**Comment by kkraus14** [comment]
> `_make_finalizer` is kept as downstream projects (cuDF and cuGRAPH) still use it, but is no longer used by RMM.

We have the new `_DevicePointer` class that uses Cython to manage lifetime of a raw pointer (will move to a smart pointer in the future), that we can use to remove those `_make_finalizer` usages.

**Comment by kkraus14** [comment]
> Alternatively, consider that C++ has a rmm::device_vector class that I think would be a more natural backing for device_array.

I think this is just a step in the direction of getting rid of the array functionality in RMM altogether. Once cuDF / cuML / cuGraph no longer depend on using `rmm.device_array` I believe the plan is to rip it out.

**Comment by jrhemstad** [comment]
> Interesting. Thanks for mentioning that. What are the main differences between the two?

`device_buffer`
- Untyped
- Uninitialized

`device_vector<T>`
- Typed
- Default initialized

**Comment by jakirkham** [comment]
Thanks for the review Keith and info Jake!

Agree about `_DevicePointer`. Should we update downstream libraries before making the change here?

Leaving things uninitialized ends up being useful in some cases (like [creating an empty array]( https://docs-cupy.chainer.org/en/stable/reference/generated/cupy.empty.html )). After all we have options to initialize these in Python later.

Is there any other effect to having typed data? Does RMM use different pools for different sized elements for instance? If not, it's probably fine for us to leave this out for now too. We can type things in Python later too.

File: python/rmm/rmm.py [hunk 1]
```diff
@@ -179,17 +170,15 @@ def device_array(shape, dtype=np.float, strides=None, order="C", stream=0):
         shape, strides, dtype.itemsize
     )
 
-    addr = librmm.rmm_alloc(datasize, stream)
+    buf = librmm.DeviceBuffer(size=datasize, stream=stream)
 
     # Note Numba will call the finalizer to free the device memory
```
Comment (kkraus14) [inline python/rmm/rmm.py]: Dead comment

**Comment by jrhemstad** [comment]
> Is there any other effect to having typed data? Does RMM use different pools for different sized elements for instance?

In a dynamically typed language like Python, the typing probably doesn't matter. In C++, having a typed storage just adds a layer of safety and explicitness that usually comes with static typing.

As far as RMM is concerned, there's no difference between the memory allocate by `device_buffer` or `device_array`.

**Pull request merged**